plugins/auth-backend-module-oidc-proxy-provider/src/helpers.ts

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2024 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { LoggerService } from '@backstage/backend-plugin-api';
+import { AuthenticationError } from '@backstage/errors';
+import { JWK, KeyLike, importJWK, jwtVerify } from 'jose';
+import fetch from 'node-fetch';
+import { IDTokenInfo } from './types';
+
+export function createTokenValidator(
+  logger: LoggerService,
+  iss: string,
+  aud: string,
+  providedJwks?: (KeyLike | Uint8Array)[],
+): (token: string) => Promise<IDTokenInfo> {
+  let jwksUri: string = '';
+
+  return async function tokenValidator(token) {
+    // Perform discovery once to avoid two round trips on every authentication request.
+    if (jwksUri === '' && providedJwks === undefined) {
+      jwksUri = await discoverJwksUri(iss);
+    }
+
+    logger.info(`jwksUri: ${jwksUri}`);
+    // Verify the token was signed by the issuer.  Performs one round trip to
+    // the jwks uri every authentication request to fetch the current key set
+    // from the issuer.  May be optimized in the future, but this happens only
+    // at sign-in, not for every request to the backend.
```

Comment:
sure, but once for every user, and in an org with thousands of people it may start to get noticeable

Reply:
This is a reflection of your comment in the gcp-iap provider, which behaves the same:  [// TODO(freben): Rate limit the public key reads. It may be sensible to](https://github.com/jeffmccune/backstage/blob/305475c1b374cae3302fdfdad97c1be296e13af5/plugins/auth-backend-module-gcp-iap-provider/src/helpers.ts#L28-L32)

Let me know if you'd like it changed, and what to, otherwise I'll leave it.